core/src/main/java/org/apache/calcite/rel/rules/ProjectCorrelateTransposeRule.java

```diff
@@ -0,0 +1,123 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Correlate;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.RelFactories;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.BitSets;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import java.util.BitSet;
+
+/**
+ * Push Project under Correlate to apply on Correlate's left and right child
+ */
+public class ProjectCorrelateTransposeRule  extends RelOptRule {
+
+  /**
+   * preserveExprCondition to define the condition for a expression not to be pushed
+   */
+  private final PushProjector.ExprCondition preserveExprCondition;
+
+  //~ Instance fields --------------------------------------------------------
+
+  public static final ProjectCorrelateTransposeRule INSTANCE =
+      new ProjectCorrelateTransposeRule(
+          PushProjector.ExprCondition.TRUE,
+          RelFactories.LOGICAL_BUILDER);
+
+  //~ Constructors -----------------------------------------------------------
+
+  public ProjectCorrelateTransposeRule(
+      PushProjector.ExprCondition preserveExprCondition,
+      RelBuilderFactory relFactory) {
+    super(
+        operand(Project.class,
+            operand(Correlate.class, any())),
+        relFactory, null);
+    this.preserveExprCondition = preserveExprCondition;
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  public void onMatch(RelOptRuleCall call) {
+    Project origProj = call.rel(0);
+    final Correlate corr = call.rel(1);
+
+    // locate all fields referenced in the projection and join condition;
+    // determine which inputs are referenced in the projection and
+    // join condition; if all fields are being referenced and there are no
+    // special expressions, no point in proceeding any further
+    PushProjector pushProject =
+        new PushProjector(
+            origProj,
+            call.builder().literal(true),
+            corr,
+            preserveExprCondition,
+            call.builder());
+    if (pushProject.locateAllRefs()) {
+      return;
+    }
+
+    // create left and right projections, projecting only those
+    // fields referenced on each side
+    RelNode leftProjRel =
+        pushProject.createProjectRefsAndExprs(
+            corr.getLeft(),
+            true,
+            false);
+    RelNode rightProjRel =
+        pushProject.createProjectRefsAndExprs(
+            corr.getRight(),
+            true,
+            true);
+
+    // convert the join condition to reference the projected columns
```

Comment:
The comment says 'convert the join condition'  but for Correlate since there is no join filter (that a regular Join has),  it is basically adjusting the correlation columns to point to the new child.  You could make that clear in the comment.